### PR TITLE
Add BackendHdf5

### PR DESCRIPTION
**Pull request by jotsetung**
With this also the `BackendHdf5` is implemented. Open things we still might want to discuss:

1) Which type of checksum are we calculating? Are we storing the checksum within the file (my favorite) *vs* calculating/defining it on the fly? (issue #429)
2) Subsetting of `Backend` classes: `backendSubset` (my favorite, in the pull request) *vs* `[,Backend` and `filterFile,Backend` (issue #417)

Note: I will not add more code to this branch until we have settled the above points.

### CODE REVIEW

File: R/Backend-class.R [hunk 1]
```diff
@@ -102,29 +117,18 @@ setValidity("Backend", function(object) {
...
-    signature="Backend",
-    definition=function(object) {
+setMethod("show", signature = "Backend", definition = function(object) {
     cat("Backend:", class(object)[1L], "\n")
-    cat("Source files:\n",
-        paste(" ", basename(object@files), collapse="\n"), "\n", sep=""
-    )
+    fls <- basename(object@files)
+    if (length(fls) > 3)
+        fls <- c(fls[1:3], paste0("(", length(fls) -3,
```
[the remark this answers is not in the document]
Comment (jotsetung) [reply R/Backend-class.R]: 😄 👍

File: R/BackendHdf5-class.R [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+#'
+#' @author Johannes Rainer
+#'
+#' @md
+#'
+#' @noRd
+setMethod("backendInitialize", "BackendHdf5", function(object, files,
+                                                       spectraData, path = ".",
+                                                       ...) {
+    path <- suppressWarnings(normalizePath(path))
```
[the remark this answers is not in the document]
Comment (jotsetung) [reply R/BackendHdf5-class.R]: OK, will fix.

File: R/BackendHdf5-class.R [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+#' @author Johannes Rainer, Sebastian Gibb
+#'
+#' @noRd
+.serialize_msfile_to_hdf5 <- function(file, h5file) {
+    h5 <- H5Fopen(h5file)
+    on.exit(H5Fclose(h5))
+    comp_level <- .hdf5_compression_level()
+    fh <- openMSfile(file)
+    hdr <- header(fh)
+    pks <- peaks(fh)
```
[the remark this answers is not in the document]
Comment (jotsetung) [reply R/BackendHdf5-class.R]: The reason is speed. Calling `peaks` repeatedly is slower than calling it once. I can change that if you want, but I don't think memory is a problem - if it is (i.e. a single mzML files is so big the data can not be kept in memory) the user will have big troubles anyway.

File: R/BackendHdf5-class.R [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+    on.exit(invisible(H5Fclose(h5)))
+    comp_level <- .hdf5_compression_level()
+    if (prune) {
+        h5delete(h5, "spectra")
+    }
+    if (!H5Lexists(h5, "spectra"))
+        h5createGroup(h5, "spectra")
+    spids <- paste0("/spectra/", spectraData$spIdx)
+    for (i in seq_along(x)) {
+        h5write(cbind(mz = mz(x[[i]]), intensity = intensity(x[[i]])),
```
[the remark this answers is not in the document]
Comment (jotsetung) [reply R/BackendHdf5-class.R]: OK.

File: R/MSnExperiment-class.R [hunk 5]
```diff
@@ -326,11 +402,60 @@ setMethod("metadata", "MSnExperiment",
...
+    if (!missing(j))
+        stop("Subsetting by columns/samples is not supported")
+    if (missing(i))
+        return(x)
+    i <- .to_index(rownames(x@spectraData), i)
+    x@spectraData <- x@spectraData[i, , drop = FALSE]
+    file <- unique(x@spectraData$fileIdx) # here we allow unsorted file idx.
+    x@spectraData$fileIdx <- match(x@spectraData$fileIdx, file)
+    x@sampleData <- x@sampleData[file, , drop = FALSE]
+    x@backend <- backendSubset(x@backend, i, file)
```
[the remark this answers is not in the document]
Comment (jotsetung) [reply R/MSnExperiment-class.R]: I will rewrite the `backendSubset` method to take argument `spectraData` to subset. After that we obviously will have to update the `$fileIdx`. Not completely sure if I got all comments from you - but I guess that is what you suggested in #417.

File: R/MSnExperiment-class.R [hunk 6]
```diff
@@ -326,11 +402,60 @@ setMethod("metadata", "MSnExperiment",
...
+          function(x, i, j = "missing", drop = "missing") {
+              x[i, , drop = TRUE]
+          })
+
+#' @rdname MSnExperiment
+setMethod("filterFile", "MSnExperiment", function(object, file) {
+    if (missing(file))
+        return(object)
+    file <- .to_index(fileNames(object), file, variable = "file")
+    names_orig <- rownames(object@spectraData)
```
[the remark this answers is not in the document]
Comment (jotsetung) [reply R/MSnExperiment-class.R]: I think that has to be handled differently anyway if we use `backendSubset` with the `spectraData` argument.

File: R/BackendHdf5-class.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (jotsetung) [reply R/BackendHdf5-class.R]: Fixed.

File: R/BackendHdf5-class.R (hunk 4 above)
[the remark this answers is not in the document]
Comment (jotsetung) [reply R/BackendHdf5-class.R]: Fixed.

File: R/BackendMemory-class.R [hunk 7]
```diff
@@ -56,33 +56,48 @@ setValidity("BackendMemory", function(object) {
...
-    validObject(object)
-    object
-})
+setMethod("backendSubset", signature(object = "BackendMemory", i = "numeric",
+                                     file = "numeric"),
+          function(object, i, file) {
+              files_orig <- object@files
+              object@files <- object@files[file]
+              ## Update also `@fromFile` in the spectra.
+              object@spectra <- lapply(object@spectra[i], function(z) {
```
[the remark this answers is not in the document]
Comment (jotsetung) [reply R/BackendMemory-class.R]: Changed/updated in the new `backendSubset` implementation.

File: R/Backend-class.R [hunk 8]
```diff
@@ -270,3 +274,27 @@ setGeneric(
         standardGeneric("backendWriteSpectra"),
     valueClass="Backend"
 )
+
+#' Subset the `Backend` by spectra (`i`) and/or file (`file`).
```
Comment (sgibb) [inline R/Backend-class.R]: This documentation is now outdated. Could you please adapt it to the new implementation.

File: R/BackendHdf5-class.R [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+#'
+#' @noRd
+setMethod("backendInitialize", "BackendHdf5", function(object, files,
+                                                       spectraData, path = ".",
+                                                       ...) {
+    path <- normalizePath(path, mustWork = FALSE)
+    dir.create(path, showWarnings = FALSE, recursive = TRUE)
+    n_files <- length(files)
+    object@files <- files
+    names(object@files) <- sprintf(
```
Comment (sgibb) [inline R/BackendHdf5-class.R]: I wanted the names for the `file` slot for easier subsetting in the `BackendMemory`. Is it now useful at all? If not you can remove it here (and I will removed it from `Backend` and `BackendMemory`).

Comment (jotsetung) [reply R/BackendHdf5-class.R]: Well, it is not used anymore, so we can remove it I guess.

File: R/Backend-class.R (hunk 8 above)
Comment (jotsetung) [reply R/Backend-class.R]: Will do.

File: R/BackendHdf5-class.R (hunk 9 above)
Comment (jotsetung) [reply R/BackendHdf5-class.R]: I've removed the code also from the `Backend` and `BackendMemory` classes to fix the unit tests. You should have a look after merging if it's OK and do some cleanup (I mostly just commented out the stuff but didn't delete).

File: R/Backend-class.R (hunk 8 above)
Comment (jotsetung) [reply R/Backend-class.R]: Done

**Pull request merged**